src/server/game/Movement/MotionMaster.cpp

```diff
@@ -640,72 +978,113 @@ void MotionMaster::MovePath(uint32 pathId, bool repeatable)
 {
     if (IsInvalidMovementGeneratorType(type))
     {
-        TC_LOG_DEBUG("movement.motionmaster", "MotionMaster::LaunchMoveSpline: '%s', tried to launch a spline with an invalid MovementGeneratorType: %u (Id: %u, Slot: %u)", _owner->GetGUID().ToString().c_str(), type, id, slot);
+        TC_LOG_DEBUG("movement.motionmaster", "MotionMaster::LaunchMoveSpline: '%s', tried to launch a spline with an invalid MovementGeneratorType: %u (Id: %u, Priority: %u)", _owner->GetGUID().ToString().c_str(), type, id, priority);
         return;
     }
 
-    TC_LOG_DEBUG("movement.motionmaster", "MotionMaster::LaunchMoveSpline: '%s', initiates spline Id: %u (Type: %u, Slot: %u)", _owner->GetGUID().ToString().c_str(), id, type, slot);
-    Mutate(new GenericMovementGenerator(std::move(init), type, id), slot);
+    TC_LOG_DEBUG("movement.motionmaster", "MotionMaster::LaunchMoveSpline: '%s', initiates spline Id: %u (Type: %u, Priority: %u)", _owner->GetGUID().ToString().c_str(), id, type, priority);
+
+    GenericMovementGenerator* movement = new GenericMovementGenerator(std::move(init), type, id);
+    movement->Priority = priority;
+    Add(movement);
 }
 
 /******************** Private methods ********************/
 
-void MotionMaster::pop()
+void MotionMaster::Pop(bool active, bool movementInform)
 {
-    if (empty())
-        return;
+    MovementGenerator* pointer = *_generators.begin();
+    _generators.erase(pointer);
+    Delete(pointer, active, movementInform);
+}
 
-    _slot[_top] = nullptr;
-    while (!empty() && !top())
-        --_top;
+void MotionMaster::DirectInitialize()
+{
+    // Clear ALL movement generators (including default)
+    DirectClearDefault();
+    DirectClear();
+
+    InitializeDefault();
 }
 
-bool MotionMaster::NeedInitTop() const
+void MotionMaster::DirectClear()
 {
-    if (empty())
-        return false;
-    return _initialize[_top];
+    // First delete Top
+    if (!_generators.empty())
+        Pop(true, false);
+
+    // Then the rest
+    while (!_generators.empty())
+        Pop(false, false);
+
+    // Make sure the storage is empty
+    ClearBaseUnitStates();
 }
 
-void MotionMaster::InitTop()
+void MotionMaster::DirectClearDefault()
 {
-    top()->Initialize(_owner);
-    _initialize[_top] = false;
+    if (_defaultGenerator)
+        DeleteDefault(_generators.empty(), false);
 }
 
-void MotionMaster::Mutate(MovementGenerator *m, MovementSlot slot)
+void MotionMaster::DirectClear(std::function<bool(MovementGenerator*)> const& filter)
 {
+    if (_generators.empty())
+        return;
+
+    MovementGenerator const* top = GetCurrentMovementGenerator();
+
+    std::vector<MovementGenerator*> cache;
+    std::for_each(_generators.begin(), _generators.end(), [filter, &cache](MovementGenerator* a)
+    {
+        if (filter(a))
+            cache.push_back(a);
+    });
+
+    if (cache.empty())
+        return;
+
+    for (MovementGenerator* movement : cache)
```

Comment:
you can do this without a temporary container, just use iterator directly instead of this loop syntax

Reply:
I guess, since the container size wont ever be > 20 (and thats maybe too much to say), I thought this was """"""safer"""""" and the setback was minimal. Will change though 👍